include/nlohmann/json.hpp

```diff
@@ -1227,13 +1233,70 @@ class basic_json
     invariant. Furthermore, it has to be called each time the type of a JSON
     value is changed, because the invariant expresses a relationship between
     @a m_type and @a m_value.
+
+    Furthermore, the parent relation is checked for arrays and objects: If
+    @a check_parents true and the value is an array or object, then the
+    container's elements must have the current value as parent.
+
+    @param[in] check_parents  whether the parent relation should be checked.
+               The value is true by default and should only be set to true
+               during destruction of objects when the invariant does not
+               need to hold.
     */
-    void assert_invariant() const noexcept
+    void assert_invariant(bool check_parents = true) const noexcept
     {
         JSON_ASSERT(m_type != value_t::object || m_value.object != nullptr);
         JSON_ASSERT(m_type != value_t::array || m_value.array != nullptr);
         JSON_ASSERT(m_type != value_t::string || m_value.string != nullptr);
         JSON_ASSERT(m_type != value_t::binary || m_value.binary != nullptr);
+
+#if JSON_DIAGNOSTICS
+        JSON_ASSERT(!check_parents || !is_structured() || std::all_of(begin(), end(), [this](const basic_json & j)
+        {
+            return j.m_parent == this;
+        }));
+#endif
+    }
+
+    reference set_parent(reference j, bool recursive)
```

Comment:
I'm concerned that this form allows you to accidentally do `j.set_parent(j, false)` and set `j`'s `m_parent` to itself.  It's also very weird to see `j.set_parent(j...` in the code.  Could these be two separate functions, such as `set_parent(reference j)` and `set_parents()`?  This way it's clear that one is setting the parent of the passed in object, and the other is setting the parents within `this`.

Reply:
Good idea. Done in b9d3aa40670a57cedeb71e52c192a7678f0e23c2.